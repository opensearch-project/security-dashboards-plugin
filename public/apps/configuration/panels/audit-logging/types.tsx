/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export interface GeneralSettings {
  // layer
  enable_rest?: boolean;
  enable_transport?: boolean;
  disabled_rest_categories?: string[];
  disabled_transport_categories?: string[];

  // attribute
  resolve_bulk_requests?: boolean;
  resolve_indices?: boolean;
  log_request_body?: boolean;
  exclude_sensitive_headers?: boolean;

  // ignore
  ignore_users?: string[];
  ignore_requests?: string[];
}

export interface ComplianceSettings {
  enabled?: boolean;

  // read
  read_metadata_only?: boolean;
  read_ignore_users?: string[];
  read_watched_fields?: {
    [key: string]: string;
  };

  // write
  write_metadata_only?: boolean;
  write_log_diffs?: boolean;
  write_ignore_users?: string[];
  write_watched_indices?: string[];
}

export interface AuditLoggingSettings {
  enabled?: boolean;
  audit?: GeneralSettings;
  compliance?: ComplianceSettings;
}
