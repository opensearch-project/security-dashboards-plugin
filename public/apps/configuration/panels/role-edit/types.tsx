/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export type ComboBoxOptions = EuiComboBoxOptionOption[];
export type FieldLevelSecurityMethod = 'exclude' | 'include';
export interface RoleIndexPermissionStateClass {
  indexPatterns: ComboBoxOptions;
  docLevelSecurity: string;
  fieldLevelSecurityMethod: FieldLevelSecurityMethod;
  fieldLevelSecurityFields: ComboBoxOptions;
  maskedFields: ComboBoxOptions;
  allowedActions: ComboBoxOptions;
}

export enum TenantPermissionType {
  None = '',
  Read = 'r',
  Write = 'w',
  Full = 'rw',
}

export interface RoleTenantPermissionStateClass {
  tenantPatterns: ComboBoxOptions;
  permissionType: TenantPermissionType;
}
