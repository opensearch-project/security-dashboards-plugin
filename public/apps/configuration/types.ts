/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiComboBoxOptionOption } from '@elastic/eui';

export type ComboBoxOptions = EuiComboBoxOptionOption[];

export type FieldLevelSecurityMethod = 'exclude' | 'include';

export enum ResourceType {
  roles = 'roles',
  users = 'users',
  permissions = 'permissions',
  tenants = 'tenants',
  tenantsManageTab = 'tenantsManageTab',
  tenantsConfigureTab = 'tenantsConfigureTab',
  auth = 'auth',
  auditLogging = 'auditLogging',
}

export enum Action {
  view = 'view',
  create = 'create',
  edit = 'edit',
  duplicate = 'duplicate',
  manage = 'manage',
  configure = 'configure',
}

export enum SubAction {
  mapuser = 'mapuser',
}

export enum PageId {
  dashboardId = 'dashboards',
  visualizationId = 'visualize',
}

export enum TenantPermissionType {
  None = '',
  Read = 'Read only',
  ReadWrite = 'Read and Write',
}

export interface RouteItem {
  name: string;
  href: string;
}

export interface DataObject<T> {
  [key: string]: T;
}

export interface ObjectsMessage<T> {
  total: number;
  data: DataObject<T>;
}

export interface RoleIndexPermission {
  index_patterns: string[];
  dls: string;
  fls: string[];
  masked_fields: string[];
  allowed_actions: string[];
}

export interface RoleIndexPermissionView extends RoleIndexPermission {
  id: number;
}

export interface RoleTenantPatterns {
  tenant_patterns: string[];
}

export interface RoleTenantPermission extends RoleTenantPatterns {
  allowed_actions: string[];
}

export interface RoleTenantPermissionView extends RoleTenantPatterns {
  permissionType: string;
}

export interface RoleUpdate {
  cluster_permissions: string[];
  index_permissions: RoleIndexPermission[];
  tenant_permissions: RoleTenantPermission[];
}

export interface RoleDetail extends RoleUpdate {
  reserved: boolean;
}

export interface RoleMappingDetail {
  backend_roles: string[];
  hosts: string[];
  users: string[];
}

export interface TenantUpdate {
  description: string;
}

export interface TenantName {
  tenant: string;
}

export interface Tenant extends TenantUpdate, TenantName {
  reserved: boolean;
  tenantValue: string;
}

export interface TenantSelect extends TenantName {
  username: string;
}

export interface RoleTenantPermissionDetail extends RoleTenantPermissionView, Tenant {}

export interface UserAttributes {
  [key: string]: string;
}

export interface InternalUser {
  attributes: UserAttributes;
  backend_roles: string[];
}

export interface InternalUserUpdate extends InternalUser {
  password?: string;
}

export interface ActionGroupUpdate {
  allowed_actions: string[];
}

export interface ActionGroupItem extends ActionGroupUpdate {
  reserved: boolean;
  type: 'cluster' | 'index' | 'all' | 'kibana' | undefined;
}

export interface ExpandedRowMapInterface {
  [key: string]: React.ReactNode;
}

export interface FormRowDeps {
  headerText: string;
  optional?: boolean;
  headerSubText?: string;
  helpLink?: string;
  helpText?: string;
}
