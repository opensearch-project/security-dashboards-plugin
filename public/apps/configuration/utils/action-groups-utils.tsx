/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { map } from 'lodash';
import { API_ENDPOINT_ACTIONGROUPS, CLUSTER_PERMISSIONS, INDEX_PERMISSIONS } from '../constants';
import { DataObject, ActionGroupItem, ActionGroupUpdate, ObjectsMessage } from '../types';
import { httpDelete, httpGet, httpPost } from './request-utils';
import { getResourceUrl } from './resource-utils';

export interface PermissionListingItem {
  name: string;
  type: 'Action group' | 'Single permission';
  reserved: boolean;
  allowedActions: string[];
  hasClusterPermission: boolean;
  hasIndexPermission: boolean;
}

export async function fetchActionGroups(http: HttpStart): Promise<DataObject<ActionGroupItem>> {
  const actiongroups = await httpGet<ObjectsMessage<ActionGroupItem>>({
    http,
    url: API_ENDPOINT_ACTIONGROUPS,
  });
  return actiongroups.data;
}

export function transformActionGroupsToListingFormat(
  rawData: DataObject<ActionGroupItem>
): PermissionListingItem[] {
  return map(rawData, (value: ActionGroupItem, key?: string) => ({
    name: key || '',
    type: 'Action group',
    reserved: value.reserved,
    allowedActions: value.allowed_actions,
    hasClusterPermission: value.type === 'cluster' || value.type === 'all',
    hasIndexPermission: value.type === 'index' || value.type === 'all',
  }));
}

export async function fetchActionGroupListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(await fetchActionGroups(http));
}

function getClusterSinglePermissions(): PermissionListingItem[] {
  return CLUSTER_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: true,
    hasIndexPermission: false,
  }));
}

function getIndexSinglePermissions(): PermissionListingItem[] {
  return INDEX_PERMISSIONS.map((permission) => ({
    name: permission,
    type: 'Single permission',
    reserved: true,
    allowedActions: [],
    hasClusterPermission: false,
    hasIndexPermission: true,
  }));
}

export async function getAllPermissionsListing(http: HttpStart): Promise<PermissionListingItem[]> {
  return mergeAllPermissions(await fetchActionGroups(http));
}

export async function mergeAllPermissions(
  actionGroups: DataObject<ActionGroupItem>
): Promise<PermissionListingItem[]> {
  return transformActionGroupsToListingFormat(actionGroups)
    .concat(getClusterSinglePermissions())
    .concat(getIndexSinglePermissions());
}

export async function updateActionGroup(
  http: HttpStart,
  groupName: string,
  updateObject: ActionGroupUpdate
): Promise<ActionGroupUpdate> {
  return await httpPost({
    http,
    url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, groupName),
    body: updateObject,
  });
}

export async function requestDeleteActionGroups(http: HttpStart, groups: string[]) {
  for (const group of groups) {
    await httpDelete({ http, url: getResourceUrl(API_ENDPOINT_ACTIONGROUPS, group) });
  }
}
