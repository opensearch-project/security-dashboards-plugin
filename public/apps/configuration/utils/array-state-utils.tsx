/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { get, set, curry, PropertyName } from 'lodash';
import { Dispatch, SetStateAction } from 'react';

function resolveValue<T>(value: T | (() => T)) {
  return typeof value === 'function' ? (value as Function)() : value;
}

/**
 * Update an element in an sub array
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern"
 * @param newValue value to be updated
 *
 * e.g.
 *  currentState = [{a: 1}, {a: 2}]
 *  path = [0, 'a']
 *  newValue = 10
 * The newState will be [{a: 10}, {a: 2}]
 */
export function updateElementInArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  setStateCallback((prevState) => {
    const newState = [...prevState];
    set(newState, path, resolveValue(newValue));
    return newState;
  });
}

export const updateElementInArrayHandler = curry(updateElementInArray);

/**
 * Append an element to an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param newValue value to be updated
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  newValue = 3
 * The newState will be [1, 2, 3]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  newValue = 5
 * The newState will be [{a: [1, 2, 5]}, {a: [3, 4]}]
 */
export function appendElementToArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  newValue: T | (() => T)
) {
  const resolvedNewValue = resolveValue(newValue);
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      return [...prevState, resolvedNewValue];
    } else {
      const newArray = [...(get(prevState, path) as T[]), resolvedNewValue];
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}

/**
 * Remove an element from an array (or sub array indicating by path)
 * @param setStateCallback setState function
 * @param path lodash path, e.g. [0, indexPattern] or "[0].indexPattern", use [] to indicate root level
 * @param index index of element to be removed
 *
 * e.g.
 * Scenario 1, path = [] to append to root level array
 *  currentState = [1, 2]
 *  path = []
 *  index = 0
 * The newState will be [2]
 *
 * Scenario 2, path != [] to append to sub array
 *  currentState = [{a: [1, 2]}, {a: [3, 4]}]
 *  path = [0, 'a']
 *  index = 1
 * The newState will be [{a: [1]}, {a: [3, 4]}]
 */
export function removeElementFromArray<T>(
  setStateCallback: Dispatch<SetStateAction<any[]>>,
  path: PropertyName | PropertyName[],
  index: number
) {
  setStateCallback((prevState) => {
    if ((path as PropertyName[]).length === 0) {
      const newState = [...prevState];
      newState.splice(index, 1);
      return newState;
    } else {
      const newArray = [...(get(prevState, path) as T[])];
      newArray.splice(index, 1);
      const newState = [...prevState];
      set(newState, path, newArray);
      return newState;
    }
  });
}
