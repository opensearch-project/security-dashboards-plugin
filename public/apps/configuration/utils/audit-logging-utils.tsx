/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { AuditLoggingSettings } from '../panels/audit-logging/types';
import { API_ENDPOINT_AUDITLOGGING, API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import { httpGet, httpPost } from './request-utils';

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await httpPost({ http, url: API_ENDPOINT_AUDITLOGGING_UPDATE, body: updateObject });
}

export async function getAuditLogging(http: HttpStart): Promise<AuditLoggingSettings> {
  const rawConfiguration = await httpGet<any>({ http, url: API_ENDPOINT_AUDITLOGGING });
  return rawConfiguration?.config;
}
