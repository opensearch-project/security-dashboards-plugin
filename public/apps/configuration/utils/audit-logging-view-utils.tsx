/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */
import {
  EuiDescribedFormGroup,
  EuiFlexGrid,
  EuiForm,
  EuiFormRow,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiSwitch,
  EuiText,
  EuiTitle,
} from '@elastic/eui';

import React from 'react';
import { HttpStart } from 'kibana/public';
import { displayArray, displayBoolean, displayObject, renderTextFlexItem } from './display-utils';
import { API_ENDPOINT_AUDITLOGGING_UPDATE } from '../constants';
import {
  AuditLoggingSettings,
  ComplianceSettings,
  GeneralSettings,
} from '../panels/audit-logging/types';

export function renderStatusPanel(onSwitchChange: () => void, auditLoggingEnabled: boolean) {
  return (
    <EuiPanel>
      <EuiTitle>
        <h3>Audit logging</h3>
      </EuiTitle>
      <EuiHorizontalRule />
      <EuiForm>
        <EuiDescribedFormGroup
          title={<h3>Enable audit logging</h3>}
          description={<>Enable or disable audit logging</>}
        >
          <EuiFormRow>
            <EuiSwitch
              name="auditLoggingEnabledSwitch"
              label={displayBoolean(auditLoggingEnabled)}
              checked={auditLoggingEnabled}
              onChange={onSwitchChange}
            />
          </EuiFormRow>
        </EuiDescribedFormGroup>
      </EuiForm>
    </EuiPanel>
  );
}

export function renderGeneralSettings(generalSettings: GeneralSettings) {
  return (
    <>
      {renderGeneralSettingsForLayer(generalSettings)}

      <EuiSpacer />

      {renderGeneralSettingsForAttribute(generalSettings)}

      <EuiSpacer />

      {renderGeneralSettingsForIgnore(generalSettings)}
    </>
  );
}

function renderGeneralSettingsForLayer(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Layer settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('REST layer', displayBoolean(generalSettings.enable_rest))}
        {renderTextFlexItem('Transport layer', displayBoolean(generalSettings.enable_transport))}
        {renderTextFlexItem(
          'REST disabled categories',
          displayArray(generalSettings.disabled_rest_categories)
        )}
        {renderTextFlexItem(
          'Transport disabled categories',
          displayArray(generalSettings.disabled_transport_categories)
        )}
      </EuiFlexGrid>
    </>
  );
}

function renderGeneralSettingsForAttribute(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Attribute settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('Bulk requests', displayBoolean(generalSettings.resolve_bulk_requests))}
        {renderTextFlexItem('Resolve indices', displayBoolean(generalSettings.resolve_indices))}
        {renderTextFlexItem('Request body', displayBoolean(generalSettings.log_request_body))}
        {renderTextFlexItem(
          'Sensitive headers',
          displayBoolean(generalSettings.exclude_sensitive_headers)
        )}
      </EuiFlexGrid>
    </>
  );
}

function renderGeneralSettingsForIgnore(generalSettings: GeneralSettings) {
  return (
    <>
      <EuiText>
        <h3>Ignore settings</h3>
      </EuiText>

      <EuiSpacer />

      <EuiFlexGrid columns={3}>
        {renderTextFlexItem('Exclude users', displayArray(generalSettings.ignore_users))}
        {renderTextFlexItem('Exclude requests', displayArray(generalSettings.ignore_requests))}
      </EuiFlexGrid>
    </>
  );
}

export function renderComplianceSettings(complianceSettings: ComplianceSettings) {
  return (
    <>
      {renderTextFlexItem('Compliance mode', displayBoolean(complianceSettings.enabled))}

      <EuiSpacer />

      <EuiPanel>
        <EuiTitle>
          <h4>Read</h4>
        </EuiTitle>

        <EuiSpacer />

        <EuiFlexGrid columns={3}>
          {renderTextFlexItem(
            'Read metadata only',
            displayBoolean(complianceSettings.read_metadata_only)
          )}
          {renderTextFlexItem('Ignored users', displayArray(complianceSettings.read_ignore_users))}
          {/* read_watched_fields is object instead of array.*/}
          {renderTextFlexItem(
            'Watched fields',
            displayObject(complianceSettings.read_watched_fields)
          )}
        </EuiFlexGrid>
      </EuiPanel>

      <EuiSpacer />

      <EuiPanel>
        <EuiTitle>
          <h4>Write</h4>
        </EuiTitle>

        <EuiSpacer />

        <EuiFlexGrid columns={3}>
          {renderTextFlexItem(
            'Write metadata only',
            displayBoolean(complianceSettings.write_metadata_only)
          )}
          {renderTextFlexItem('Log diffs', displayBoolean(complianceSettings.write_log_diffs))}
          {renderTextFlexItem('Ignored users', displayArray(complianceSettings.write_ignore_users))}
          {renderTextFlexItem(
            'Watched indices',
            displayArray(complianceSettings.write_watched_indices)
          )}
        </EuiFlexGrid>
      </EuiPanel>
    </>
  );
}

export async function updateAuditLogging(http: HttpStart, updateObject: AuditLoggingSettings) {
  return await http.post(`${API_ENDPOINT_AUDITLOGGING_UPDATE}`, {
    body: JSON.stringify(updateObject),
  });
}
