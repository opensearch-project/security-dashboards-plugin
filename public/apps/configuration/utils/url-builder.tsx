/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { Action } from '../types';
import { ResourceType } from '../../../../common';

/**
 * Build hash based url, encode the resourceId part
 * e.g.
 *   buildUrl() => "/" (landing page)
 *   buildUrl(ResourceType.roles) => "/roles" (role listing page)
 *   buildUrl(ResourceType.roles, Action.create) => "/roles/create" (role creation page)
 *   buildUrl(ResourceType.roles, Action.view, "../someRole") => "/roles/view/..%2FsomeRole" (role detail page)
 *   buildUrl(ResourceType.roles, Action.view, "someRole", SubAction.mapuser) => "/roles/view/someRole/mapuser" (Map User page)
 * edge case (wrong usage) fallbacks:
 *   buildUrl(undefined, Action.create) => ""
 *   buildUrl(ResourceType.roles, undefined, "someRole") => "/roles"
 */
export function buildUrl(
  resouceType?: ResourceType,
  action?: Action,
  resourceId?: string,
  subAction?: string
) {
  const encodedResouceId = resourceId ? encodeURIComponent(resourceId) : undefined;
  const rawContents = [resouceType, action, encodedResouceId, subAction];
  const contents = [];
  for (const content of rawContents) {
    if (content === undefined) {
      break;
    }
    contents.push(content);
  }
  return '/' + contents.join('/');
}

export function buildHashUrl(
  resouceType?: ResourceType,
  action?: Action,
  resourceId?: string,
  subAction?: string
) {
  return '#' + buildUrl(resouceType, action, resourceId, subAction);
}
