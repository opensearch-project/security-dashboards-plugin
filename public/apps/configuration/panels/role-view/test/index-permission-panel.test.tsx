/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { Action, RoleIndexPermissionView } from '../../../types';
import { ResourceType } from '../../../../../../common';
import {
  renderFieldLevelSecurity,
  renderRowExpanstionArrow,
  toggleRowDetails,
  IndexPermissionPanel,
  renderDocumentLevelSecurity,
} from '../index-permission-panel';
import { EuiButtonIcon, EuiEmptyPrompt, EuiInMemoryTable } from '@elastic/eui';
import { buildHashUrl } from '../../../utils/url-builder';

describe('Role view - index permission panel', () => {
  const sampleRoleName = 'role1';
  const sampleRoleIndexPermission: RoleIndexPermissionView = {
    id: 1,
    index_patterns: [],
    dls: '',
    fls: [],
    masked_fields: [],
    allowed_actions: [],
  };
  const setMap = jest.fn();

  it('Toggle row details', () => {
    toggleRowDetails(sampleRoleIndexPermission, {}, setMap);
    const updateMapFunc = setMap.mock.calls[0][0];
    updateMapFunc({ group: '' });
  });

  describe('Render row expanstion arrow', () => {
    it('should render down arrow when collapsed', () => {
      const renderFunc = renderRowExpanstionArrow({}, {}, jest.fn());
      const Wrapper = () => <>{renderFunc(sampleRoleIndexPermission)}</>;
      const component = shallow(<Wrapper />);

      expect(component.find(EuiButtonIcon).prop('iconType')).toBe('arrowDown');
    });

    it('should render up arrow when expanded', () => {
      const renderFunc = renderRowExpanstionArrow(
        { [sampleRoleIndexPermission.id]: sampleRoleIndexPermission },
        {},
        jest.fn()
      );
      const Wrapper = () => <>{renderFunc(sampleRoleIndexPermission)}</>;
      const component = shallow(<Wrapper />);

      expect(component.find(EuiButtonIcon).prop('iconType')).toBe('arrowUp');
    });

    it('renders when arrow expanded', () => {
      const renderFunc = renderRowExpanstionArrow(
        { [sampleRoleIndexPermission.id]: sampleRoleIndexPermission },
        {},
        jest.fn()
      );
      const Wrapper = () => <>{renderFunc(sampleRoleIndexPermission)}</>;
      const component = shallow(<Wrapper />);
      component.find(EuiButtonIcon).simulate('click');
      expect(component).toMatchSnapshot();
    });
  });

  describe('Render field level security', () => {
    const field = 'fls';
    const excludeField = '~' + field;

    it('renders em-dash when field level security is empty', () => {
      const renderFunc = renderFieldLevelSecurity();
      const Wrapper = () => <>{renderFunc([])}</>;
      const component = mount(<Wrapper />);
      expect(component.find('[data-test-subj="empty-fls-text"]').first().text()).toEqual('â€”');
    });

    it('renders field level security when it contains exclude fields', () => {
      const renderFunc = renderFieldLevelSecurity();
      const Wrapper = () => <>{renderFunc([excludeField])}</>;
      const component = mount(<Wrapper />);
      expect(component.find('[data-test-subj="fls-text"]').first().text()).toEqual(
        'Exclude: ' + field
      );
    });

    it('renders field level security when it contains include fields', () => {
      const renderFunc = renderFieldLevelSecurity();
      const Wrapper = () => <>{renderFunc([field])}</>;
      const component = mount(<Wrapper />);
      expect(component.find('[data-test-subj="fls-text"]').first().text()).toEqual(
        'Include: ' + field
      );
    });
  });

  describe('Render document level security', () => {
    it('renders em-dash when document level security is empty', () => {
      const renderFunc = renderDocumentLevelSecurity();
      const Wrapper = () => <>{renderFunc('')}</>;
      const component = shallow(<Wrapper />);
      expect(component).toMatchSnapshot();
    });

    it('renders document level security when dls is valid json', () => {
      const renderFunc = renderDocumentLevelSecurity();
      const Wrapper = () => <>{renderFunc('{"key": "value"}')}</>;
      const component = shallow(<Wrapper />);
      expect(component).toMatchSnapshot();
    });

    it('renders document level security when dls is invalid json', () => {
      const renderFunc = renderDocumentLevelSecurity();
      const Wrapper = () => <>{renderFunc('{"key": "value"%%%}')}</>;
      const spy = jest.spyOn(console, 'warn').mockImplementationOnce(() => {});
      const component = shallow(<Wrapper />);
      expect(spy).toBeCalled();
      expect(component).toMatchSnapshot();
    });
  });

  describe('Index permission list', () => {
    it('render empty', () => {
      const wrapper = shallow(
        <IndexPermissionPanel
          roleName={sampleRoleName}
          indexPermissions={[]}
          actionGroups={{}}
          errorFlag={false}
          loading={false}
          isReserved={false}
        />
      );

      expect(wrapper.find(EuiInMemoryTable).prop('items').length).toBe(0);
    });

    it('it renders a empty prompt when indexPermissions is empty', () => {
      const wrapper = mount(
        <IndexPermissionPanel
          roleName={sampleRoleName}
          indexPermissions={[]}
          actionGroups={{}}
          errorFlag={false}
          loading={false}
          isReserved={false}
        />
      );
      const prompt = wrapper
        .find('[data-test-subj="index-permission-container"] tbody EuiEmptyPrompt')
        .first()
        .getElement();
      expect(prompt.type).toBe(EuiEmptyPrompt);
    });

    it('Add index permission button is disabled for reserved role', () => {
      const wrapper = mount(
        <IndexPermissionPanel
          roleName={sampleRoleName}
          indexPermissions={[]}
          actionGroups={{}}
          errorFlag={false}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="index-permission-container"] tbody EuiEmptyPrompt')
        .first()
        .getElement();
      const component = shallow(prompt);
      const button = component.find('[data-test-subj="addIndexPermission"]');
      expect(button.prop('disabled')).toBe(true);
    });

    it('should render to role edit page when click on Add index permission button', () => {
      const wrapper = mount(
        <IndexPermissionPanel
          roleName={sampleRoleName}
          indexPermissions={[]}
          actionGroups={{}}
          errorFlag={false}
          loading={false}
          isReserved={true}
        />
      );

      const prompt = wrapper
        .find('[data-test-subj="index-permission-container"] tbody EuiEmptyPrompt')
        .first()
        .getElement();
      const component = shallow(prompt);
      component.find('[data-test-subj="addIndexPermission"]').simulate('click');
      expect(window.location.hash).toBe(
        buildHashUrl(ResourceType.roles, Action.edit, sampleRoleName)
      );
    });
  });
});
