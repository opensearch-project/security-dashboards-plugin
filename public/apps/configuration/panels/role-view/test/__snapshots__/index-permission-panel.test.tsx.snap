// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Role view - index permission panel Render document level security renders document level security when dls is invalid json 1`] = `
<Fragment>
  {"key": "value"%%%}
</Fragment>
`;

exports[`Role view - index permission panel Render document level security renders document level security when dls is valid json 1`] = `
<Fragment>
  <ExpressionModal
    expression={
      Object {
        "key": "value",
      }
    }
    title="Document-level security"
  />
</Fragment>
`;

exports[`Role view - index permission panel Render document level security renders em-dash when document level security is empty 1`] = `
<Fragment>
  <span>
    â€”
  </span>
</Fragment>
`;

exports[`Role view - index permission panel Render row expanstion arrow renders when arrow expanded 1`] = `
<Fragment>
  <EuiButtonIcon
    aria-label="Collapse"
    iconType="arrowUp"
    onClick={[Function]}
  />
</Fragment>
`;
